engine/execution/computation/computer/uploader/uploader_test.go

```diff
@@ -95,35 +99,75 @@ func Test_AsyncUploader(t *testing.T) {
 		require.Equal(t, 3, callCount)
 	})
 
-	time.Sleep(1 * time.Second)
-
+	// This test shuts down the async uploader right after the upload has started. The upload has an error to force
+	// the retry mechanism to kick in (under normal circumstances). Since the component is shutting down, the retry
+	// should not kick in.
+	//
+	// sequence of events:
+	// 1. create async uploader and initiate upload with an error - to force retrying
+	// 2. shut down async uploader right after upload initiated (not completed)
+	// 3. assert that upload called only once even when trying to use retry mechanism
 	t.Run("stopping component stops retrying", func(t *testing.T) {
 		testutils.SkipUnless(t, testutils.TEST_FLAKY, "flaky")
 
 		callCount := 0
+		t.Log("test started grID:", string(bytes.Fields(debug.Stack())[1]))
 
-		wg := sync.WaitGroup{}
-		wg.Add(1)
+		// this wait group ensures that async uploader has a chance to start the upload before component is shut down
+		// otherwise, there's a race condition that can happen where the component can shut down before the async uploader
+		// has a chance to start the upload
+		wgUploadStarted := sync.WaitGroup{}
+		wgUploadStarted.Add(1)
+
+		// this wait group ensures that async uploader won't send an error (to test if retry will kick in) until
+		// the component has initiated shutting down (which should stop retry from working)
+		wgShutdownStarted := sync.WaitGroup{}
+		wgShutdownStarted.Add(1)
+		t.Log("added 1 to wait group grID:", string(bytes.Fields(debug.Stack())[1]))
 
 		uploader := &DummyUploader{
 			f: func() error {
-				defer func() {
+				t.Log("DummyUploader func() - about to call wgUploadStarted.Done() grID:", string(bytes.Fields(debug.Stack())[1]))
```

Comment:
I'm not sure we should use `debug.Stack()` on normal operation. I get it helps, but seems like potential pitfall.

Reply:
It's strictly for debugging purposes to log the goroutine id - it was very helpful in debugging this test and I think it's something we should use for tests with multiple goroutines. What's the potential pitfall?

Reply:
If there's a better way to log the goroutine ID, let's use that, but this is the most convenient way I found.